### ISSUE TITLE
Opcode Search Tool Enhancements

### ISSUE THREAD

**Issue by OrionSR** [sannybuilder/dev#76]
What started with an idea to provide View options for the Opcode Search Tool, balloon into some major upgrades as I read through the documentation and worked up a mockup image. My original thought was that users probably won't be happy if I change the commands files too much and take away a more familiar reference. 

![EnhancedOpcodeSearchTool](https://user-images.githubusercontent.com/15115889/94330084-a43dee00-ff75-11ea-8bce-f78aaf0c346b.png)
**Enhancements:** 

Multiple views of the command examples - bottom
- Opcodes - Traditional examples that show all opcodes
- Keywords - Same as above but decompiled without the Write Opcodes option
- In-Game - Either taken from main.scm or a variation with realistic data values - numbers not vars, general locals for handles, no arrays.
- Advanced - Either SCM.ini (debugging) or the examples from the edit modes with the other parameter order.

Tool Bar - left
- Hotkey for managing opcodes as button icons for easy access
- Insert added, selected opcode is inserted directly into the script

Operators - top center
- Buttons insert the operators into the search field at the cursor

Saved Searches - top right
- Pre-defined searches that group commands by purpose or type
- Selectable from a drop menu, or whatever else works
- Ability to add and remove custom searches

**Comment by MatiDragon-YT** [comment]
Hey The idea is not bad at all, although... that interface seems not very professional.
Things seem to be a bit messy.
The things that would be used most often are somewhat far from others, and the less used things are among the most used...
And I have a question.
How would the `saved` ones be deleted?
I hope you are not offended by my comment.

> ### post data:
> #### here is a more organized design that you may like.
> ![search](https://user-images.githubusercontent.com/43966706/94356944-b2216b00-006a-11eb-8092-83a17a224583.png)

**Comment by OrionSR** [comment]
> that interface seems not very professional
> I hope you are not offended by my comment
Yes, I agree. And no, I'm not insulted. I've never been happy with the quality of my artistry but have learned to be proud of how my drawings communicate complex ideas.

What is the purpose of your slider?

I started looking for icons but gave up on Insert, and decided the crappy text buttons communicated the purpose more clearly. What I imagined though were familiar buttons for common editing tasks, with hover text for additional help and the hotkey that can be used instead. I have a hard time remembering the hotkeys and wanted to make them more accessible and easier to learn.

I put editing buttons on the left near the opcode list because that's what they are supposed to operate on.
What you labeled as Filters would more accurately be described as Operators, and defined in the help documentation.

I originally used Filters as the label for Searches, but I'd expect a proper filter to apply regardless of the search string.  Saved searches would need to be named, and that would be the key for deletion. Not really sure how to prevent the user from deleting the standard filters though. The other goal of the Searches was to provide examples of complex search strings.

I would be sad if the Views were not more easily accessible as this was the primary inspiration for the idea. What I had imaged were tabs but, like I mentioned, graphics are not my forte. I had already spent more time on the image that I wanted.

What the image couldn't communicate was not losing my place as I switched views. This might be difficult with the keyword examples without opcodes to key on. Maybe the tool can just trim the first 6 characters from the opcode.txt (examples.txt) to provide the type of example one would expect when using high-level constructs or with Write Opcodes disabled.

LIT, for literal, is intended to allow for searches like underscore space, or period space. There's a lot of things this tool can't search for but I don't see a clear solution. Figured I'd just throw it in with the batch though.

Clear, btw, would be more appropriate as an X at the end of the search string box, but that bit is covered by the Searches box in both of our images.

**Comment by MatiDragon-YT** [comment]
oh.. well, I had the idea that both you and I could customize it a bit like this:

**Tool Bar**
* Ability to be dragged as the upper bar of Sanny Builder.

**Icons**
* Display floating text by hovering over it for a short time.  Like the help messages in the program settings section.

**Slider**
* well, regarding this one .. I just didn't want to leave the `sort` option so alone, so I thought .. why not control the size of the letters?  and I left the `F` for Fonts to make its function more notable. Simply put, it's filler :)

Equally your work is very well done, since it serves a lot to prove if things fulfill the main objective. And if all conditions are fulfilled, then you could work in the design to more thoroughly.

**Comment by OrionSR** [comment]
> Oh.. well, I had the idea that both you and I could customize it a bit like this:
I doubt I can be of much help with actually coding anything. I'm a bit overwhelmed with the prospect of completing the command overhaul.  At my current rate of process it will take many months to complete the overhaul on each game and apply similar changes to SB's class based param order. 

I suppose an opcode search tool could be developed independently as a user tool. I think Seemann had other ideas for an opcode search tool, like moving it to an online version. My concern with an online tool though, is that it wouldn't operate on my current file configuration. 

I like idea of a draggable tool bar. especially if it can be dragged to a vertical position next to the opcode list.

I tend to have lots on ideas as I'm working on Sanny projects; some useful, others not so much. So I plant them as seeds to see what grows into something beautiful, or useful to more people than just myself.

**Comment by MatiDragon-YT** [comment]
Oh, don't question yourself so much sir.
We know that you have a brilliant mind and always do the right thing for others.
Often times I also get a lot of ideas while working on my projects and making them, but I leave them as hidden functions or as separate download tools. And if they are well received, I implement them in the final program.
Perhaps you can do the same.
Perhaps you can do the same with some of your ideas :smile:

**Comment by wmysterio** [comment]
Buttons clutter up the interface. Maybe need group the filters to combobox and put them with other filters to main panel. It if filters will display correctly when draggable panel.
![94356944-b2216b00-006a-11eb-8092-83a17a224583](https://user-images.githubusercontent.com/27928887/94415996-f43fcf00-0186-11eb-841b-1e72db47af99.png)

**Comment by MatiDragon-YT** [comment]
Ehm... I think I ran into some problems with your design.
* The optimization with the time that is wasted in selecting the "filters".
* Seeing that every time you open the tool, the search engine is already selected so that you can write, I think that the groups that you defined would mean that you lose some time, although it would be great, if instead of looking for Cars, Actors or Objects, it would select the CREATE, IF, SET, GET and ADD.
* The ability to adapt vertically, that if it can be included, your design could take up a lot of space in the window. This because you left everything in the same bar.

**Comment by OrionSR** [comment]
I must admit though, I really like the clean design of the original interface. The interest in this issue has generated a few more ideas, but I don't think I want to embarrass myself with another unprofessional mockup. ;)

**Operators:** IF AND OR NOT Sequential, and hopefully Literal. 
- These are being described inaccurately as Filters. They _operate_ on the search parameters, and the term _Filters_ is needed for the filter groups. 
-The idea of the Operator Bar was inspired by how the github editor simply inserts the proper characters needed for special formatting.
- To simplify the interface design, the operators could be displayed as hover text for the search box.

**Filters** as proper group filters. Abandon the idea of saved or predefined searches.
- Show All at the top. with radio buttons to enable or disable specific groups.
- Filters would be independent of the search box. Users could search within the enabled groups.
- Instead of saved searches, allow the user to save Favorite opcodes for specific projects.
- Filters would be defined in a text file with a title followed by a list of opcode numbers.

**Sort and Hide Opcodes** as one button toggles.
- The sort button could toggle between 123 and ABC.
- Hide could simply grey out when enabled.
- Hiding the opcodes could maintain the numeric sort order.
- I anticipate issues with Hide on Sanny's class based opcodes.

**Editing Toolbar** as a pop-out side panel?

**Views** as a toggle between a generic and descriptive example text, and an in-game variant.
- Genaric: `00C0: set_time_of_day $Hours $Minutes`
- In-Game:  `00C0: set_time_of_day 6 30`

**NEW Descriptions** as hover text for each opcode.
- From Opcode Database `00C0: Sets the current in-game _clock_ to the time specified in the two parameters.`
- PRINTF() formatting to allow for linefeeds used to display special parameter values like; Use -1 for random models

**NEW Title Bar** at the very top, where the filename is displayed in the Sanny Editor
- File name of the current root mode (modes for modes.xml)
- Title of the current edit mode  (GTA San Andreas v1)

Much of the inspiration behind this design is drawn from the information provided in the [Opcode Database](https://gtagmodding.com/opcode-database/)

Um... there was something else. I'll update when I remember.

**Comment by OrionSR** [comment]
Ah, here we go...

Right-Clicking an opcode entry displays only the Copy function. Including the other toolbar options (Copy All, Add, Insert_ on this menu would be intuitive and wouldn't clutter the interface. I suppose Sort and Hide option might fit naturally on this menu as well. Also Help, which can link to external resources.

Um... I was going to suggest that Operators could be added to a R-Click menu of the search box, but this menu is already pretty long, and includes a lot of options I don't understand.
![OpcodeSearchNotes](https://user-images.githubusercontent.com/15115889/94493016-6c66be80-01a0-11eb-9022-01645b2d8b6d.png)

**Comment by MiranDMC** [comment]
All propositions seems to be extreme total overhauls.

For now I propose to add:
*Support of Ctrl+C shortcut
*Support of Ctrl+V shortcut
*Double click on list would insert selected list into editor window at caret position as new line.

**Comment by x87** [comment]
> Double click on list would insert selected row into editor window at caret position as new line.

I think Enter should do it. 


My other ideas:

- Use F1 to switch between the editor and open OST window
- When the cursor is on line like 
```
06E1: task_car_mission char [Char] vehicle [Car] targetVehicle [Car] missionId [CarMission] cruiseSpeed [float] drivingStyle [DrivingMode]
```
clicking Tab moves the cursor to the next square bracketed word ([Char], [Car]) and select it, so you can easily replace it, then click Tab, then replace next and so cycle through the argument placeholders.
